src/libsystemd-network/dhcp-option.c

```diff
@@ -97,30 +97,92 @@ static int option_append(uint8_t options[], size_t size, size_t *offset,
                 *offset += 2;
 
                 ORDERED_SET_FOREACH(p, s) {
-                        options[*offset] = p->option;
-                        options[*offset + 1] = p->length;
-                        memcpy(&options[*offset + 2], p->data, p->length);
-                        *offset += 2 + p->length;
+                        //Not checking error code(destination size already checked)
+                        dhcp_option_append_tlv(options, size, offset, p->option, p->length, p->data);
                 }
 
                 break;
         }
+        case SD_DHCP_OPTION_RELAY_AGENT_INFORMATION: {
+                sd_dhcp_server *server = (sd_dhcp_server *) optval;
+                if (optlen > 255)
+                        return log_dhcp_server_errno(server, SYNTHETIC_ERRNO(ENOBUFS), "relay agent circuit and/or remote id is too long: %m");
+                options[(*offset)++] = code;
+                options[(*offset)++] = optlen;
+
+                if (server->agent_circuit_id) {
+                        size_t length = strlen(server->agent_circuit_id);
+                        int subcode = SD_DHCP_RELAY_AGENT_CIRCUIT_ID;
+                        int r = dhcp_option_append_tlv(options, size, offset, subcode, length, server->agent_circuit_id);
+                        if (r < 0)
+                                return r;
+                }
+                if (server->agent_remote_id) {
+                        size_t length = strlen(server->agent_remote_id);
+                        int subcode = SD_DHCP_RELAY_AGENT_REMOTE_ID;
+                        int r = dhcp_option_append_tlv(options, size, offset, subcode, length, server->agent_remote_id);
+                        if (r < 0)
+                                return r;
+                }
+                break;
+
+        }
         default:
-                if (*offset + 2 + optlen > size)
-                        return -ENOBUFS;
+                return dhcp_option_append_tlv(options, size, offset, code, optlen, optval);
+        }
 
-                options[*offset] = code;
-                options[*offset + 1] = optlen;
+        return 0;
+}
+
+/* Append type-length value structure to the options buffer */
+int dhcp_option_append_tlv(uint8_t options[], size_t size, size_t *offset, uint8_t code, size_t optlen, const void *optval) {
```

Comment:
It is not necessary to expose this function. Please make this `static` and move it before `option_append()`.

Reply:
Thank you guys for the review! I have addressed the issues, and pushed the changes. Ready for another review.